lib/actions/config.js

```diff
@@ -0,0 +1,417 @@
+
+Examples:
+
+	$ balena config generate --device 7cf02a6 --version 2.12.7
+	$ balena config generate --device 7cf02a6 --version 2.12.7 --generate-device-api-key
+	$ balena config generate --device 7cf02a6 --version 2.12.7 --device-api-key <existingDeviceKey>
+	$ balena config generate --device 7cf02a6 --version 2.12.7 --output config.json
+	$ balena config generate --app MyApp --version 2.12.7
+	$ balena config generate --app MyApp --version 2.12.7 --device-type fincm3
+	$ balena config generate --app MyApp --version 2.12.7 --output config.json
+	$ balena config generate --app MyApp --version 2.12.7 \
+--network wifi --wifiSsid mySsid --wifiKey abcdefgh --appUpdatePollInterval 1\
+`,
+	options: [
+		commandOptions.osVersion,
+		commandOptions.optionalApplication,
+		commandOptions.optionalDevice,
+		commandOptions.optionalDeviceApiKey,
+		commandOptions.optionalDeviceType,
+		{
+			signature: 'generate-device-api-key',
+			description: 'generate a fresh device key for the device',
+			boolean: true,
+		},
+		{
+			signature: 'output',
+			description: 'output',
+			parameter: 'output',
+			alias: 'o',
+		},
+		// Options for non-interactive configuration
+		{
+			signature: 'network',
+			description: 'the network type to use: ethernet or wifi',
+			parameter: 'network',
+		},
+		{
+			signature: 'wifiSsid',
+			description:
+				'the wifi ssid to use (used only if --network is set to wifi)',
+			parameter: 'wifiSsid',
+		},
+		{
+			signature: 'wifiKey',
+			description:
+				'the wifi key to use (used only if --network is set to wifi)',
+			parameter: 'wifiKey',
+		},
+		{
+			signature: 'appUpdatePollInterval',
+			description:
+				'how frequently (in minutes) to poll for application updates',
+			parameter: 'appUpdatePollInterval',
+		},
+	],
+	permission: 'user',
+	action(_params, options) {
+		normalizeUuidProp(options, 'device');
+		const Promise = require('bluebird');
+		const writeFileAsync = Promise.promisify(require('fs').writeFile);
+		const balena = getBalenaSdk();
+		const form = require('resin-cli-form');
+		const prettyjson = require('prettyjson');
+
+		const {
+			generateDeviceConfig,
+			generateApplicationConfig,
+		} = require('../utils/config');
+		const helpers = require('../utils/helpers');
+		const { exitWithExpectedError } = require('../utils/patterns');
+
+		if (options.device == null && options.application == null) {
+			exitWithExpectedError(`\
+You have to pass either a device or an application.
+
+See the help page for examples:
+
+  $ balena help config generate\
+`);
+		}
+
+		if (!options.application && options.deviceType) {
+			exitWithExpectedError(`\
+Specifying a different device type is only supported when
+generating a config for an application:
+
+* An application, with --app <appname>
+* A specific device type, with --device-type <deviceTypeSlug>
+
+See the help page for examples:
+
+  $ balena help config generate\
+`);
+		}
+
+		return Promise.try(
+			/** @returns {Promise<any>} */ function() {
```

Comment:
Had to add this to cast the return type, `Bluebird.try` isn't very happy with inferring the return type when it can be different things based on the branch